src/kOS.Safe/Screen/ScreenBuffer.cs

```diff
@@ -70,15 +85,45 @@ protected void AddNewBufferLines( int howMany = 1)
                 buffer.Add(new char[ColumnCount]);
         }
 
-        public void SetSize(int rowCount, int columnCount)
+        public void AddResizeNotifier(ScreenBuffer.ResizeNotifier notifier)
+        {
+            if (notifyees.IndexOf(notifier) < 0)
+                notifyees.Add(notifier);
+        }
+
+        public void RemoveResizeNotifier(ScreenBuffer.ResizeNotifier notifier)
+        {
+            notifyees.Remove(notifier);
+        }
+
+        public void SetSize(int rows, int columns)
+        {
+            RowCount = rows;
+            ColumnCount = columns;
+            ResizeBuffer();
+            int scrollDiff = 0;
+            foreach (ResizeNotifier notifier in notifyees)
+            {
+                if (notifier != null)
+                    scrollDiff += notifier(this);
+            }
+            ScrollVertical(scrollDiff);
+        }
+        
+        protected void ResizeBuffer()
         {
-            if (rowCount <= MAX_ROWS && columnCount <= MAX_COLUMNS)
+            // Grow or shrink the width of the buffer lines to match the new
+            // value.  Note that this does not (yet) account for preserving lines and wrapping them.
+            for (int row = 0 ; row < buffer.Count ; ++row)
             {
-                RowCount = rowCount;
-                ColumnCount = columnCount;
-                buffer.Clear();
-                InitializeBuffer();
+                char[] newRow = new char[ColumnCount];
+                Array.Copy(buffer[row], 0, newRow, 0, Math.Min(buffer[row].Length, ColumnCount));
+                buffer[row] = newRow;
             }
+            
+            // Add more buffer lines if needed to pad out the rest of the screen:
+            while (buffer.Count - topRow < RowCount)
+                buffer.Add(new char[ColumnCount]);
         }
```

Comment:
And this is how it preserves the contents.  When the terminal size changes. it copies the old data into the new buffer its making, row for row, cutting off old content that was past the end (if shrinking the width or height).

For this first pass, the old content is thrown away when cropped off the new smaller size.  For a more sophisticated future pass it may be possible to wrap long lines when the width is shortened, and unwrap them when it's widened, but doing so requires that the ScreenBuffer record the difference between a "soft" end of line wrap and "hard" return, so it knows where the wrapping happens.  Right now that information is lost after the Print() statement is done.